content-security-policy/navigation/javascript-url-navigation-to-child-snapshots-parent-csp.html

```diff
@@ -0,0 +1,69 @@
+<!DOCTYPE html>
+<html>
+<head>
+  <meta charset="utf-8">
+  <script src="/resources/testharness.js"></script>
+  <script src="/resources/testharnessreport.js"></script>
+</head>
+<body>
+  <iframe id="iframe"></iframe>
+  <script nonce="abc">
+    // The intention of this test is to ensure a snapshot of the CSP from the
+    // time of creating the `javascript:` URL navigation tasks is used when
+    // executing the tasks.
+    // The spec is buggy, see
+    // <https://github.com/whatwg/html/issues/4651#issuecomment-2412623188>
+    // and the encompassing ticket. This is one step towards correcting it.
```

Comment:
I suggest adding an outline of the queued tasks somewhere, as I had to write it out myself to determine that 2 was the correct number of calls to f(). Something like:

```
On load event:
- Queue task 1 for call-1 to f
- Queue task 2 for call to addCSPWithFrameSrcNonceAbcToDoc
- Queue task 3 for call-2 to f

Task 1:
  - Call-1 to f happens (successfully)
Task 2:
  - Add CSP
  - Queue task 4 to call g
Task 3:
  - Call-2 to f happens (successfully)
Task 4:
  - Call to g happens (failure)
  - Queue task 5 for securitypolicyviolation
Task 5:
  - Final asserts and end test
```

However, writing this out, I don't think this test checks the snapshotting behavior? The desired snapshotting behavior is checking during step 2's "queue task 4 to call g". Whereas the current buggy spec behavior is checking during step 4. But per the above sequence, the CSP is present at both points.

Unless I'm missing something (very possible), I think the more accurate test would change task 2 to flip "add CSP" and "queue task 4 to call g", and then assert that g succeeds.

Reply:
> Whereas the current buggy spec behavior is checking during step 4. But per the above sequence, the CSP is present at both points.

The CSP is also present during task 3, which is required to succeed. If it would check the present CSP (instead of the snapshotted) during task execution, it'd fail.

> I think the more accurate test would change task 2 to flip "add CSP" and "queue task 4 to call g", and then assert that g succeeds.

That would require another test checks against any (present or snapshotted) CSP during task execution. Presumably there is such test. If that's clearer to you, I'll change it accordingly.

Reply:
> The desired snapshotting behavior is checking during step 2's "queue task 4 to call g"

Yes. Still need to think about your proposal and potentially figure out another way to actually test the desired behavior.

Reply:
> Unless I'm missing something (very possible), I think the more accurate test would change task 2 to flip "add CSP" and "queue task 4 to call g", and then assert that g succeeds.

With that change:
- `g` could only fail if it checks the present CSP, not if it checks the snapshotted one from the time of creating task 4.
- call-2 to `f` could only fail if it checks the present CSP, not if it checks the snapshotted one from the time of creating task 3.

So unless I misunderstood, a different secenario for testing the snapshotted CSP is checked at the time of task creation and not during task execution is needed.

Reply:
@domenic: rewrote the test, including more detailed explanations.